token-erc-721/chaincode-javascript/lib/tokenERC721.js

```diff
@@ -0,0 +1,484 @@
+    */
+    async TokenURI(ctx, tokenId) {
+        const nft = await this._readNFT(ctx, tokenId);
+        return nft.tokenURI;
+    }
+
+    // ============== ERC721 enumeration extension ===============
+
+    /**
+     * TotalSupply counts non-fungible tokens tracked by this contract.
+     *
+     * @param {Context} ctx the transaction context
+     * @returns {Number} Returns a count of valid non-fungible tokens tracked by this contract,
+     * where each one of them has an assigned and queryable owner.
+     */
+    async TotalSupply(ctx) {
+        const tokensAsBytes = await ctx.stub.getState(tokenArrayKey);
+        const tokens = JSON.parse(tokensAsBytes.toString());
+        return tokens.length;
+    }
+
+    /**
+     * TokenByIndex enumerates valid non-fungible tokens.
+     *
+     * @param {Context} ctx the transaction context
+     * @param {Number} index A counter less than `totalSupply()`
+     * @returns {Number} Returns the tokenId for the index'th non-fungible token, (sort order not specified)
+     */
+    async TokenByIndex(ctx, index) {
+        const tokensAsBytes = await ctx.stub.getState(tokenArrayKey);
+        const tokens = JSON.parse(tokensAsBytes.toString());
+
+        const indexInt = parseInt(index);
+        if (indexInt >= tokens.length) {
+            throw new Error('The index is invalid. Index must be less than totalSupply()');
+        }
+
+        return tokens[indexInt];
+    }
+
+    /**
+     * Enumerate non-fungible tokens assigned to an owner.
+     *
+     * @param {Context} ctx the transaction context
+     * @param {String} owner token owner
+     * @param {Number} index A counter less than `balanceOf(owner)`
+     * @returns {Number} Returns the tokenId for the index'th non-fungible token assigned to the owner, (sort order not specified)
+     */
+    async TokenOfOwnerByIndex(ctx, owner, index) {
+        const balance = await this._readBalance(ctx, owner);
+        const indexInt = parseInt(index);
+        if (indexInt >= balance.tokens.length) {
+            throw new Error('The index is invalid. Index must be less than balanceOf(owner)');
+        }
+
+        return balance.tokens[indexInt];
+    }
+
+
+    // ============== Extended Functions for this sample ===============
+
+    /**
+     * Set optional information for a token.
+     *
+     * @param {Context} ctx the transaction context
+     * @param {String} name The name of the token
+     * @param {String} symbol The symbol of the token
+     */
+    async SetOption(ctx, name, symbol) {
+        await ctx.stub.putState(nameKey, Buffer.from(name));
+        await ctx.stub.putState(symbolKey, Buffer.from(symbol));
+        return true;
+    }
+
+    /**
+     * Mint a new non-fungible token
+     *
+     * @param {Context} ctx the transaction context
+     * @param {String} tokenId Unique ID of the non-fungible token to be minted
+     * @returns {Object} Return the non-fungible token object
+    */
+    async Mint(ctx, tokenId) {
```

Comment:
Have you considered allowing a user to mint a new token and set the token URI in a single call, rather than two calls (mint + set token URI)?

Reply:
The OpenZeppelin implementation also seems to have a `mintWithTokenURI`: https://docs.openzeppelin.com/contracts/2.x/api/token/erc721#ERC721MetadataMintable-mintWithTokenURI-address-uint256-string-

Reply:
Thanks. I missed that part. It seems we have two options: 1) Use separated functions `Mint` + `SetTokenURI`, 2) Use a single function `MintWithTokenURI`. Which is better for this sample?

Reply:
I would prefer a single function `MintWithTokenURI`.

Reply:
Done. The updated code uses a single function `MintWithTokenURI` to mint a new token.